diff --git a/__init__package.mustache b/__init__package.mustache
index 7d1cb69..517e438 100644
--- a/__init__package.mustache
+++ b/__init__package.mustache
@@ -13,6 +13,7 @@ __all__ = [
     {{/apis}}{{/apiInfo}}"ApiResponse",
     "ApiClient",
     "Configuration",
+    "settings",
     "OpenApiException",
     "ApiTypeError",
     "ApiValueError",
diff --git a/api.mustache b/api.mustache
index 3e440e1..2817838 100644
--- a/api.mustache
+++ b/api.mustache
@@ -3,6 +3,8 @@
 {{>partial_header}}
 
 import warnings
+import json
+import re
 from pydantic import validate_call, Field, StrictFloat, StrictStr, StrictInt
 from typing import Any, Dict, List, Optional, Tuple, Union
 from typing_extensions import Annotated
@@ -14,6 +16,46 @@ from typing_extensions import Annotated
 from {{packageName}}.api_client import ApiClient, RequestSerialized
 from {{packageName}}.api_response import ApiResponse
 from {{packageName}}.rest import RESTResponseType
+from {{packageName}}.exceptions import ApiException
+
+
+def fix_file_url(url: str, token: str) -> str:
+    return re.sub(
+        r"(\/webservice)?\/pluginfile\.php", f"/tokenpluginfile.php/{token}", url
+    )
+
+def is_primitive(value: Any):
+    return isinstance(value, (int, str, bool))
+
+
+def is_collection(value: Any):
+    return isinstance(value, (set, list))
+
+
+def _parse_form(data: Any, prefix: str = "") -> List[Tuple[str, str]]:
+    if data is None:
+        return []
+    if is_primitive(data):
+        value = str(data).lower() if isinstance(data, bool) else str(data)
+        return [(prefix, value)]
+    result: List[Tuple[str, str]] = []
+    if is_collection(data):
+        for i, value in enumerate(data):
+            result = result + _parse_form(
+                value, prefix + (f"{i}" if prefix == "" else f"[{i}]")
+            )
+    else:
+        dictionary = data.__dict__ if hasattr(data, "__dict__") else data
+        for key in dictionary.keys():
+            value = dictionary[key]
+            result = result + _parse_form(
+                value, prefix + (f"{key}" if prefix == "" else f"[{key}]")
+            )
+    return result
+
+
+def parse_form(data: Dict[str, Any]):
+    return _parse_form(data)
 
 
 {{#operations}}
@@ -28,6 +70,16 @@ class {{classname}}:
         if api_client is None:
             api_client = ApiClient.get_default()
         self.api_client = api_client
+
+
+    async def get_file(self, url, userprivateaccesskey: Optional[str]=None):
+        if userprivateaccesskey is None and hasattr(self, "core_webservice_get_site_info"):
+            userprivateaccesskey = (await self.core_webservice_get_site_info()).userprivateaccesskey
+        if userprivateaccesskey is None:
+            raise ApiException("Current site is missing userprivateaccesskey")
+        return await self.api_client.call_api(
+            method="GET", url=fix_file_url(url, userprivateaccesskey)
+        )
 {{#operation}}
 
 
@@ -101,6 +153,8 @@ class {{classname}}:
             {{/allParams}}
         }
 
+        _resource_path = '{{path}}'.split('#')[0]
+
         _path_params: Dict[str, str] = {}
         _query_params: List[Tuple[str, str]] = []
         _header_params: Dict[str, Optional[str]] = _headers or {}
@@ -116,6 +170,9 @@ class {{classname}}:
             _path_params['{{baseName}}'] = {{paramName}}{{#isEnumRef}}.value{{/isEnumRef}}
 {{/pathParams}}
         # process the query parameters
+        if '/webservice/rest/server.php' in _resource_path:
+            _query_params.append(('moodlewsrestformat', 'json'))
+            _query_params.append(('wsfunction', '{{operationId}}'))
 {{#queryParams}}
         if {{paramName}} is not None:
             {{#isDateTime}}
@@ -148,6 +205,21 @@ class {{classname}}:
             _query_params.append(('{{baseName}}', {{paramName}}{{#isEnumRef}}.value{{/isEnumRef}}))
             {{/isDate}}{{/isDateTime}}
 {{/queryParams}}
+        if '/lib/ajax/service-nologin.php' in _resource_path:
+            _args_param = None
+            for _param in _query_params:
+                if _param[0] == 'args':
+                    _args_param = _param
+            if _args_param is not None:
+                _query_params.remove(_args_param)
+            else:
+                _args_param = ('args', {})
+            _new_args = ('args', json.dumps([{
+                'index': 0,
+                'methodname': '{{operationId}}',
+                'args': _args_param[1]
+            }]))
+            _query_params.append(_new_args)
         # process the header parameters
 {{#headerParams}}
         if {{paramName}} is not None:
@@ -160,7 +232,7 @@ class {{classname}}:
             _files['{{{baseName}}}'] = {{paramName}}
             {{/isFile}}
             {{^isFile}}
-            _form_params.append(('{{{baseName}}}', {{paramName}}))
+            _form_params += parse_form({'{{{baseName}}}': {{paramName}}})
             {{/isFile}}
 {{/formParams}}
         # process the body parameter
@@ -226,7 +298,7 @@ class {{classname}}:
 
         return self.api_client.param_serialize(
             method='{{httpMethod}}',
-            resource_path='{{{path}}}',
+            resource_path=_resource_path,
             path_params=_path_params,
             query_params=_query_params,
             header_params=_header_params,
diff --git a/api_client.mustache b/api_client.mustache
index e89b611..a3d058a 100644
--- a/api_client.mustache
+++ b/api_client.mustache
@@ -446,6 +446,25 @@ class ApiClient:
         if data is None:
             return None
 
+        # special case login
+        if isinstance(data, dict) and "error" in data:
+            raise ApiException(status=400, reason=data["error"], data=data)
+
+        # special case rest webservices
+        if isinstance(data, dict) and "exception" in data:
+            raise ApiException(status=400, reason=data["message"], data=data)
+
+        # special case ajax response
+        if isinstance(data, list) and len(data) == 1 and isinstance(data[0], dict) and "error" in data[0]:
+            _data = data[0]
+            if _data["error"]:
+                raise ApiException(
+                    status=400,
+                    reason=_data["exception"]["message"],
+                    data=_data
+                )
+            data = _data["data"]
+
         if isinstance(klass, str):
             if klass.startswith('List['):
                 m = re.match(r'List\[(.*)]', klass)
diff --git a/configuration.mustache b/configuration.mustache
index 2601d75..5bbd19a 100644
--- a/configuration.mustache
+++ b/configuration.mustache
@@ -20,6 +20,12 @@ import urllib3
 from {{packageName}}.signing import HttpSigningConfiguration
 {{/hasHttpSignatureMethods}}
 
+class _Settings:
+    relaxe_all_models: bool = False
+    relaxed_models: set[str] = set()
+
+settings = _Settings()
+
 JSON_SCHEMA_VALIDATION_KEYWORDS = {
     'multipleOf', 'maximum', 'exclusiveMaximum',
     'minimum', 'exclusiveMinimum', 'maxLength',
diff --git a/exports_package.mustache b/exports_package.mustache
index 96bd44e..55f0c1a 100644
--- a/exports_package.mustache
+++ b/exports_package.mustache
@@ -5,6 +5,7 @@
 from {{packageName}}.api_response import ApiResponse as ApiResponse
 from {{packageName}}.api_client import ApiClient as ApiClient
 from {{packageName}}.configuration import Configuration as Configuration
+from {{packageName}}.configuration import settings as settings
 from {{packageName}}.exceptions import OpenApiException as OpenApiException
 from {{packageName}}.exceptions import ApiTypeError as ApiTypeError
 from {{packageName}}.exceptions import ApiValueError as ApiValueError
diff --git a/model_generic.mustache b/model_generic.mustache
index 9fa35c8..e85aeb6 100644
--- a/model_generic.mustache
+++ b/model_generic.mustache
@@ -12,6 +12,8 @@ import json
 from typing import Optional, Set
 from typing_extensions import Self
 
+from {{packageName}}.configuration import settings
+
 {{#hasChildren}}
 {{#discriminator}}
 {{! If this model is a super class, importlib is used. So import the necessary modules for the type here. }}
@@ -276,8 +278,10 @@ class {{classname}}({{#parent}}{{{.}}}{{/parent}}{{^parent}}BaseModel{{/parent}}
         if obj is None:
             return None
 
+        relaxed = settings.relaxe_all_models or "{{{classname}}}" in settings.relaxed_models
+
         if not isinstance(obj, dict):
-            return cls.model_validate(obj)
+            return cls.model_construct(**obj) if relaxed else cls.model_validate(obj)
 
         {{#disallowAdditionalPropertiesIfNotPresent}}
         {{^isAdditionalPropertiesTrue}}
@@ -288,7 +292,7 @@ class {{classname}}({{#parent}}{{{.}}}{{/parent}}{{^parent}}BaseModel{{/parent}}
 
         {{/isAdditionalPropertiesTrue}}
         {{/disallowAdditionalPropertiesIfNotPresent}}
-        _obj = cls.model_validate({
+        data = {
             {{#allVars}}
             {{#isContainer}}
             {{#isArray}}
@@ -383,7 +387,8 @@ class {{classname}}({{#parent}}{{{.}}}{{/parent}}{{^parent}}BaseModel{{/parent}}
             {{/isPrimitiveType}}
             {{/isContainer}}
             {{/allVars}}
-        })
+        }
+        _obj = cls.model_construct(**data) if relaxed else cls.model_validate(data)
         {{#isAdditionalPropertiesTrue}}
         # store additional fields in additional_properties
         for _key in obj.keys():
diff --git a/pyproject.mustache b/pyproject.mustache
index d5b2826..a0c06c7 100644
--- a/pyproject.mustache
+++ b/pyproject.mustache
@@ -91,7 +91,7 @@ mypy = ">= 1.5"
 
 
 [build-system]
-requires = ["setuptools"]
+requires = ["setuptools", "wheel"]
 build-backend = "setuptools.build_meta"
 
 [tool.pylint.'MESSAGES CONTROL']
